proposals/separate-get-set-field.md

```diff
@@ -0,0 +1,187 @@
+newtype TimeOfDay = TimeOfDay
+    { seconds :: Pico
+    -- hour = seconds `div` 3600
+    -- minute = seconds `div` 60 `mod` 60
+    -- second = seconds `mod` 60
+    }
+
+newtype MyFlags = MyFlags
+    { flags :: Word8
+    -- foo = testBit flags 0
+    -- bar = testBit flags 1
+    -- baz = testBit flags 0 || testBit flags 1
+    }
+```
+
+The above types all benefit from read-only virtual fields because they store
+the internal data in a compact format, but want to be have the same ergonomics
+as types that store it in fully expanded form. They also have invariants so we
+specifically need only a virtual GetField and not a SetField.
+
+Virtual fields (but more magical):
+
+```haskell
+data FieldLenses r (f :: Type -> Type) = FieldLenses
+
+l :: FieldLenses r f
+l = FieldLenses
+
+instance (SetField x r a, Functor f) => GetField x (FieldLenses r f) ((a -> f a) -> r -> f r) where
+    getField _ = lens (getField @x) (setField @x)
+```
+
+The above when combined with the dot-syntax would allow for you to write
+`foo & l.name .~ 5` instead of `foo & l @"name" .~ 5` or `foo & l #name .~ 5`.
+
+More generally it allows you to define potentially infinite and fairly magical
+"records" which you can use clean and intuitive dot-syntax to work with.
+
+Obviously the above has absolutely no chance of supporting a reasonable
+equivalent `SetField`.
+
+## Proposed Change Specification
+
+`GHC.Records.HasField` would be adjusted from it's current planned
+specification:
+
+```haskell
+class HasField x r a | x r -> a where
+    hasField :: r -> (a -> r, a)
+```
+
+to become
+
+```haskell
+class GetField (x :: k) r a | x r -> a where
+    getField :: r -> a
+
+class GetField x r a => SetField (x :: k) r a | k r -> a where
+    setField :: a -> r -> r
+    setField = updateField @x . const
+
+    updateField :: (a -> a) -> r -> r
+    updateField f r = let (g, a) = hasField @x r in g (f a)
+
+    hasField :: r -> (a -> r, a)
+    hasField r = (\a -> setField @x a r, getField @x r)
+
+    {-# MINIMAL setField | updateField | hasField #-}
+```
+
+
+## Effect and Interactions
+
+This change would allow for the definition of various `GetField` instances
+that were mentioned above as desirable. This would then enable those instances
+to be utilized by dot-notation.
+
+This would not be backwards compatible with the existing implementation,
+but neither is the planned implementation. Unlike the planned implementation
+it would be a very mechanical rename to fix, as opposed to a potentially
+impossible definition of `hasField`.
+
+
+## Costs and Drawbacks
+
+The drawbacks seem minimal compared to the planned implementation. You can just
+treat `SetField` as the proposed `HasField` and everything will work the same.
+
+There is slightly more complexity due to having two separate classes rather
+than just one.
+
+
+## Alternatives
+
+Separate out further into "get-only", "set-only", "get-0-or-more",
+"get-and-set" etc. instead of only "get-only" and "get-and-set".
```

Comment:
How do people feel about doing something like [this snippet](https://gist.github.com/adamgundry/a7fde332b7e4db1480775a5fdad79183) or https://github.com/ghc-proposals/ghc-proposals/pull/158#issuecomment-542697634 to better support partial fields? The cost is a bit more implementation complexity including an extra type parameter on the classes, and `GetField` dictionaries no longer perfectly matching selector functions in the partial case, but the benefit is that we can define a sum-of-records and use it without the possibility of runtime crashes.

I generally agree that the `GHC.Records` classes should be tied fairly closely to Haskell's records, rather than supporting more general optics, as it is easy enough to convert into a lens/optics representation when more elaborate constructions are needed. But as partial selector functions are in Haskell 98 (and are a bit of a wart, arguably), perhaps it is worth improving their treatment here?

Reply:
I'd raise a flag, that a time example used would like to have both, `set-only` and `get-only` fields.

You can `get` day of a month from `Day`, and you can `modify` them in two ways ([clip](https://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Calendar.html#v:addGregorianMonthsClip) or [rollover](addGregorianMonthsRollOver)).

So with virtual fields IMO both read-only and modify-only are useful. There are cases (as above), where these cannot be combined into `get-and-set`, but then they should have different names: problem solved.

Note: get-only fields is not really any closer to "Haskell records" than set-only fields, IMHO.

Reply:
Personally I think sum-of-records as currently implemented is a huge wart and I am ok with not really supporting it.

I would expect any reasonable implementation of sum-of-records to support the following:

```
data FooBar
    = Foo { foo :: Int, baz :: Int }
    | Bar { bar :: Int, qux :: Int }

fooBar :: FooBar -> Int
fooBar (Foo f) = f.foo + f.baz
fooBar (Bar b) = b.bar - b.qux
```

I am hoping that long term the above will be supported once anonymous rows/records come around. So in the short term I would personally prefer not to design features that would not be compatible with this approach.

sum-of-records will still be supported via NamedFieldPuns / pattern matching, which is compatible with having overlapping labels.

Reply:
@phadej 

Could you fully explain what set-only fields you would expect for `Day`, ideally in the form of working code, because it is very much not clear to me how you would end up with a legal set-only field that couldn't be made into a get-and-set field.

Reply:
> I am hoping that long term the above will be supported once anonymous rows/records come around. So in the short term I would personally prefer not to design features that would not be compatible with this approach.

I don't think partial fields cause any additional problems here, do they? Even with a single constructor / total field there would potentially be an odd mismatch between
```hs
foobar :: Foo -> Int
foobar (Foo f) = f.foo + f.bar -- possible future

foobar f = f.foo + f.bar -- with HasField as currently envisaged
```
although in principle I think this is all solvable if we have a separate type for anonymous records, because we could imagine something like
```hs
data FooBar
    = Foo (Record { foo :: Int, baz :: Int })
    | Bar (Record { bar :: Int, qux :: Int })

foobar (Foo f) = f.foo + f.bar -- works using HasField "foo" (Record {...}) Int Total
foobar f       = f.foo + f.bar -- works using HasField "foo" FooBar Int Partial
```

Reply:
Yeah I was envisioning a separate type, so both of the first examples would work.

```
-- works using HasField "foo" FooBar Int Partial
```

Hmm I'm not so sure how that would work, since `Record` is an actual standalone type now, and thus it should be treated like any other type.

```
data FooBar
    = Foo FooType
    | Bar BarType

foobar :: FooBar -> ...
foobar f = f.foo
```

Under what circumstances would the above typecheck? You can't really do an "or" when it comes to instances, particularly not a clever "or" that gives you a different type when both match.